### PR TITLE
ZeroPadding1D and ZeroPadding3D

### LINKED ISSUE #122: Add ZeroPadding1D and ZeroPadding3D layers
Currently, the support for `ZeroPadding1D ` and `ZeroPadding3D `  are missed, and it would be great to add support for these layers. The desired PR addressing this issue should include:

- [ ]  Implementation of layer class named as `ZeroPadding1D ` (you can take inspiration from the implementation of [`ZeroPadding2D `][1] as reference)
- [ ]   Implementation of layer class named as `ZeroPadding3D ` (you can take inspiration from the implementation of [`ZeroPadding2D `][1] as reference)
- [ ] Common hierarchy of all ZeroPadding layers with abstract class with the common functionality
- [ ]  Documentation of layer and all non-private methods
- [ ]   JUnit tests in `api ` module
- [ ]   Support for export of layer to JSON (see [ModelSaver.kt][2])
- [ ]   Support for import of layer from JSON (see [ModelLoader.kt][3])

Also, if needed, you can take a look at Keras documentation for [`ZeroPadding1D `](https://keras.io/api/layers/reshaping_layers/zero_padding1d/) and [`ZeroPadding3D `](https://keras.io/api/layers/reshaping_layers/zero_padding3d/). 

**NOTE**: for the moment, there is no need to add support for "data format" (i.e., channels last vs. channels first) in your PR; you can assume the channels are always in the last dimension.

[1]: https://github.com/JetBrains/KotlinDL/blob/master/api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/reshaping/ZeroPadding2D.kt

[2]: https://github.com/JetBrains/KotlinDL/blob/master/api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/inference/keras/ModelSaver.kt

[3]: https://github.com/JetBrains/KotlinDL/blob/master/api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/inference/keras/ModelLoader.kt

P.S. If you want to take this ticket, please leave the comment below
P.P.S Read the [Contributing Guidelines](https://github.com/JetBrains/KotlinDL/blob/master/CONTRIBUTING.md).

### PR DESCRIPTION
**Pull request by therealansh**
This PR adds ZeroPadding1D and ZeroPadding3D layers. Resolves #122.

TODO:
- [x] Add implementation of `ZeroPadding1D` and `ZeroPadding3D`
- [x] Export to ModelSaver and ModelLoader
- [x] Add Docs
- [x] Add unit tests

### CODE REVIEW

File: api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/reshaping/ZeroPadding1D.kt [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+import org.tensorflow.op.Ops
+
+/**
+ * Zero-padding layer for 1D input (e.g. audio).
+ *
+ * This layer can add zeros in the rows of the audio tensor
+ *
+ * @property [padding] 2 numbers  interpreted as `(left_pad, right_pad)`.
+ *
+ * @see [Tensorflow implementation](https://github.com/tensorflow/tensorflow/blob/582c8d236cb079023657287c318ff26adb239002/tensorflow/python/keras/layers/convolutional.py#L2699)
```
Comment (zaleslaw) [inline api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/reshaping/ZeroPadding1D.kt:21]: I suggest removing the link to TF implementation

File: api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/reshaping/ZeroPadding3D.kt [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+import org.tensorflow.op.Ops
+
+/**
+ * Zero-padding layer for 3D input (e.g. video).
+ *
+ * This layer can add zeros in the rows, cols and depth of an video tensor.
+ *
+ * @property [padding] 6 numbers  interpreted as `(left_dim1_pad, right_dim1_pad, left_dim2_pad, right_dim2_pad, left_dim3_pad, right_dim3_pad)`.
+ *
+ * @see [Tensorflow implementation](https://github.com/tensorflow/tensorflow/blob/582c8d236cb079023657287c318ff26adb239002/tensorflow/python/keras/layers/convolutional.py#L2890)
```
Comment (zaleslaw) [inline api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/reshaping/ZeroPadding3D.kt:21]: Please remove the reference to the TF implementation

**Review by zaleslaw — CHANGES_REQUESTED** [review]
@therealansh So this is great PR; I want to ask you, how do you think will we have a benefit here if we create a common AbstractZeroPadding class, or it will have a tiny number of common code rows?

If you suppose that there are no benefits, please merge the master into your branch and fix two comments, otherwise, make a refactoring.

**Comment by therealansh** [comment]
Hey @zaleslaw, I think abstraction will help the project and the codebase. I'll get back with the required abstraction.

**Comment by zaleslaw** [comment]
dear @therealansh is it ready for final review? I see that abstraction was added ...
Could you please also merge master into your branch if it's finished

**Comment by therealansh** [comment]
Hey @zaleslaw , Rebased this, now its ready to be merged

File: api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/reshaping/AbstractZeroPadding.kt [hunk 3]
```diff
@@ -0,0 +1,35 @@
+package org.jetbrains.kotlinx.dl.api.core.layer.reshaping
+
+import org.jetbrains.kotlinx.dl.api.core.layer.Layer
+import org.tensorflow.Operand
+import org.tensorflow.op.Ops
+
+public abstract class AbstractZeroPadding(
```
Comment (zaleslaw) [inline api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/reshaping/AbstractZeroPadding.kt:7]: Please, add KDocs here

File: api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/reshaping/ZeroPadding1D.kt [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+import org.tensorflow.Shape
+import org.tensorflow.op.Ops
+
+/**
+ * Zero-padding layer for 1D input (e.g. audio).
+ *
+ * This layer can add zeros in the rows of the audio tensor
+ *
+ * @property [padding] 2 numbers  interpreted as `(left_pad, right_pad)`.
+ *
```
Comment (zaleslaw) [inline api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/reshaping/ZeroPadding1D.kt:18]: remove the blank line

File: api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/reshaping/ZeroPadding1D.kt [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+ */
+public class ZeroPadding1D : AbstractZeroPadding {
+    public val padding: IntArray
+    private lateinit var inputShape: Shape
+
+    /**
+    //     * Constructs an instance of ZeroPadding1D layer
```
Comment (zaleslaw) [inline api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/reshaping/ZeroPadding1D.kt:25]: This constructor has a commented KDoc, please change it

File: api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/reshaping/ZeroPadding3D.kt [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+import org.tensorflow.Shape
+import org.tensorflow.op.Ops
+
+/**
+ * Zero-padding layer for 3D input (e.g. video).
+ *
+ * This layer can add zeros in the rows, cols and depth of an video tensor.
+ *
+ * @property [padding] 6 numbers  interpreted as `(left_dim1_pad, right_dim1_pad, left_dim2_pad, right_dim2_pad, left_dim3_pad, right_dim3_pad)`.
+ *
```
Comment (zaleslaw) [inline api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/reshaping/ZeroPadding3D.kt:18]: Please remove the blank line

File: api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/reshaping/ZeroPadding3D.kt [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+ */
+public class ZeroPadding3D : AbstractZeroPadding {
+    //TODO add dataFormat support
```
Comment (zaleslaw) [inline api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/reshaping/ZeroPadding3D.kt:21]: I suggest removing TODOs related to dataFormat

File: api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/inference/keras/config/KerasPadding.kt [hunk 8]
```diff
@@ -52,4 +66,62 @@ internal sealed class KerasPadding {
             this.padding[3] = padding[1][1]
         }
     }
+
+
+    class ZeroPadding3D : KerasPadding {
```
Comment (zaleslaw) [inline api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/inference/keras/config/KerasPadding.kt:71]: Please add KDocs for this class and make it or public if it's part of API or internal if it should be available out of the library

File: api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/inference/keras/config/KerasPadding.kt [hunk 9]
```diff
@@ -52,4 +66,62 @@ internal sealed class KerasPadding {
...
+                    this.padding[4] = padding[4]
+                    this.padding[5] = padding[5]
+                }
+                else -> {
+                    throw UnsupportedOperationException("Padding with size ${padding.size} is not supported!")
+                }
+            }
+        }
+
+        constructor(padding: Array<IntArray>) {
```
Comment (zaleslaw) [inline api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/inference/keras/config/KerasPadding.kt:101]: Add KDocs

File: api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/reshaping/ZeroPadding3D.kt [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+                paddingFirstDim = intArrayOf(padding[0], padding[0])
+                paddingSecondDim = intArrayOf(padding[1], padding[1])
+                paddingThirdDim = intArrayOf(padding[2], padding[2])
+            }
+            1 -> {
+                paddingFirstDim = intArrayOf(padding[0], padding[0])
+                paddingSecondDim = intArrayOf(padding[0], padding[0])
+                paddingThirdDim = intArrayOf(padding[0], padding[0])
+            }
+            else -> throw IllegalArgumentException("Invalid padding argument at layer $name")
```
Comment (zaleslaw) [inline api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/reshaping/ZeroPadding3D.kt:119]: Add . symbol at the end of error message

**Review by zaleslaw — CHANGES_REQUESTED** [review]
This huge PR could be merged after some minor fixes in formatting, missed KDocs, visibility modifiers.

File: api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/reshaping/ZeroPadding2D.kt [hunk 11]
```diff
@@ -20,9 +18,8 @@ import org.tensorflow.op.Ops
  *
  * @property [padding] 4 numbers  interpreted as `(top_pad, bottom_pad, left_pad, right_pad)`.
  *
- * @see [Tensorflow implementation](https://github.com/tensorflow/tensorflow/blob/582c8d236cb079023657287c318ff26adb239002/tensorflow/python/keras/layers/convolutional.py#L2765)
  */
-public class ZeroPadding2D : Layer {
+public class ZeroPadding2D : AbstractZeroPadding {
     public val padding: IntArray
     private val dataFormat: String
     private lateinit var inputShape: Shape // TODO: refactor this shape to another property in another place
```
Comment (therealansh) [inline api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/reshaping/ZeroPadding2D.kt:25]: Hey @zaleslaw  should i also remove this TODO from ZeroPadding2D?

Comment (zaleslaw) [reply api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/reshaping/ZeroPadding2D.kt:22]: Please remove everywhere in this PR TODOs related to the order of the channels NWHC. This TODO looks unclear for me, explain here better or remove it.

**Comment by therealansh** [comment]
Hey @zaleslaw , i think this PR is ready now.

**Comment by zaleslaw** [comment]
It was a great contribution, many thanks @therealansh

**Pull request merged**